### PR TITLE
Print trailing comma for type parameters

### LINKED ISSUE #3662: Trailing commas for type parameters in TypeScript
**Prettier 1.9.2**

~~I would paste a playground link but no TypeScript support ;)~~ I'm silly: [Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKnAWwwBsBDeAIQEsoATGgc2wF4AdKbT7AH2wEEMJOPiJlKNelAbsuPbBVJRFpESXJxqdRjK68AwqQBOhiDFVi4AOQi047ANzt26LNgCyNuMWu2WOzrwCQh623p7+cgpKSiFePnYcutgGxqaxYbYOTlBwAB4YEIYw2JBQAM7FYIZw6gDKEACuhmBWngBiJgTm6pqSTMzYADwRuN3iWlLYefB0ZXiEauN9LPOiPRKMADQj9U0t8VO5M7Rzu82tvgNn++FQAHwAFPBrS4xIeGMaG1IAlCx32GAEQA9AAqbAAOih2FBwPYAF97CBNiAIBgYFRoGVkKAjCYAO4ABSMCGxKFIADcIFRaMiQKQKsgAGakYhlOAogBGhlIYAA1nAYLUMLzGMgYIYGhyQLQIGBmaz2SiaOyioSeQwCKQFWzpQArMq5Cg8-mC2qkAhwAAyNDgOqVqIaMAwToATPbpSLDKrkKhMHAylUqOi6RhDDQYAB1GkwAAWyAAHAAGFFhiDsyM8jC+sMBuCGCl2lHVACODSo1XVpE12qQLN1KPZBCo4sl0rKjGIcAAig1THa64rpTBSJzo7Q48hXSiJaQqMRGHoIAQtb6oNAiyAGuzcKOyfX2fD4UA)
```sh
--parser typescript
```

**Input:**
```typescript
type TemplateBinding =
    | AppleTemplateBinding
    | BananaTemplateBinding
    | CarrotTemplateNode
;

type ModelNode =
    | AppleModelNode
    | BananaModelNode
    | CarrotModelNode
;

export const createSourceNodeFromTemplateBinding = <
    TTemplateBinding extends TemplateBinding = TemplateBinding,
    TSourceNode extends SourceNode = SourceNode
>(templateBinding: TTemplateBinding) => {
    /* ... */
};
```

**Expected behavior:**

There should be a trailing comma after `= SourceNode`.

**Actual behavior:**

As of https://github.com/prettier/prettier/commit/fa708d102a6a4993d1bc548732ba1a6cda48e095, it will not.

TypeScript 2.7 will be the first version to allow it (https://github.com/Microsoft/TypeScript/issues/16152). Should prettier add them back if TS is > 2.7?

### LINKED ISSUE #9338: TypeScript trailing comma on type parameters
When setting the `--trailing-comma` setting to "all", prettier does not add a comma to TypeScript type parameters.

Example:

```typescript
export function long_long_long_long_long_long_long_function<
  AAAAAAA,
  BBBBBBB,
  CCCCCCC,
  DDDDDDD,
  EEEEEEE,  // <--- prettier removes this comma!
>(
  aaaaaa: AAAAAAA,
  bbbbbb: BBBBBBB,
  cccccc: CCCCCCC,
  dddddd: DDDDDDD,
  eeeeee: EEEEEEE,
) {
  console.log(aaaaaa, bbbbbb, cccccc, dddddd, eeeeee);
}
```

It looks like at some point in the past, prettier actually did add the trailing commas to type parameters, but this was removed in #1832 because TypeScript apparently didn't support it. TypeScript does now though (I'm not sure since which version, see <https://github.com/microsoft/TypeScript/issues/16152>)

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #3662.

I think we can do this at next major.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by thorn0** [comment]
TS 2.7 was released 3 years ago. I don't think we need a major version for this.

**Review by thorn0 — CHANGES_REQUESTED** [review]
This logic relies on the file extension (`isTSXFile`), so I can't reproduce this on the playground:

`test-tsx.tsx`:
```ts
const f = <Tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt, >() => 1
```

```sh
$ prettier test-tsx.tsx --trailing-comma all
const f = <
  Tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt,,
>() => 1;
```

See also https://github.com/prettier/prettier/issues/3662#issuecomment-530540734

BTW, this also means that Prettier has been outputting code with trailing commas in type parameters for a while already, at least for TSX. so this change is definitely safe for a non-major version.

File: src/language-js/print/type-parameters.js [hunk 1]
```diff
@@ -52,17 +52,22 @@ function printTypeParameters(path, options, print, paramsKey) {
...
+  // Keep comma if the file extension is .tsx and
+  // has one type parameter that isn't extend with any types.
+  // Because, otherwise formatted result will be invalid as tsx.
+  const trailingComma =
+    getFunctionParameters(n).length === 1 &&
+    isTSXFile(options) &&
+    !n[paramsKey][0].constraint &&
+    path.getParentNode().type === "ArrowFunctionExpression"
+      ? ","
+      : ifBreak(shouldPrintComma(options, "all") ? "," : "");
```
Comment (thorn0) [inline src/language-js/print/type-parameters.js:64]: Please extract the `shouldPrintComma` condition out of `ifBreak` like I did in `print/array.js`:

https://github.com/prettier/prettier/blob/bba719283d484f27659e2a69ff214cf287f15ab7/src/language-js/print/array.js#L60-L66

File: src/language-js/print/type-parameters.js [hunk 2]
```diff
@@ -52,17 +52,24 @@ function printTypeParameters(path, options, print, paramsKey) {
...
+  // has one type parameter that isn't extend with any types.
+  // Because, otherwise formatted result will be invalid as tsx.
+  const trailingComma =
+    getFunctionParameters(n).length === 1 &&
+    isTSXFile(options) &&
+    !n[paramsKey][0].constraint &&
+    path.getParentNode().type === "ArrowFunctionExpression"
+      ? ","
+      : shouldPrintComma(options, "all")
+      ? ifBreak(",", "")
```
Comment (thorn0) [inline src/language-js/print/type-parameters.js:65]:
```suggestion
      ? ifBreak(",")
```

**Review by thorn0 — APPROVED** [review]

**Review by fisker — APPROVED** [review]

**Pull request merged**